File: 36. Valid Sudoku/IsValidSudoku.py
'''
https://leetcode.com/problems/valid-sudoku/
'''

class Solution:
    def isValidSudoku(self, board):
        row_nums, col_nums = [set() for i in range(9)], [set() for i in range(9)]
        sub_box = {}
        for row in range(9):
            for col in range(9):
                curr_num = board[row][col]
                if curr_num == ".": continue
                if curr_num in row_nums[row] or curr_num in col_nums[col]:    return False
                row_nums[row].add(curr_num)
                col_nums[col].add(curr_num)

                key = (row//3, col//3)
                sub_box_nums = sub_box.get(key, {})
                if curr_num in sub_box_nums:    return False
                sub_box_nums.add(curr_num)
                sub_box[key] = sub_box_nums

        return True


sol = Solution()
print(sol.isValidSudoku([["5","3",".",".","7",".",".",".","."],["6",".",".","1","9","5",".",".","."],[".","9","8",".",".",".",".","6","."],["8",".",".",".","6",".",".",".","3"],["4",".",".","8",".","3",".",".","1"],["7",".",".",".","2",".",".",".","6"],[".","6",".",".",".",".","2","8","."],[".",".",".","4","1","9",".",".","5"],[".",".",".",".","8",".",".","7","9"]]))  

File: 36. Valid Sudoku/is_valid_sudoku.rs
use std::collections::HashMap;

impl Solution {
    
    // O(1) time,
    // O(1) space,
    // Approach: hashtable, 
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        let dimension = 9;
        let mut rows: Vec<HashMap<char, i32>> = vec![HashMap::new(); dimension];
        let mut cols: Vec<HashMap<char, i32>> = vec![HashMap::new(); dimension];
        
        let mut mini_squares: Vec<Vec<HashMap<char, i32>>> = vec![vec![HashMap::new();3];3];
        
        for row in 0..dimension {
            for col in 0..dimension {
                
                let val = board[row][col];
                let mut count = 0;
                if rows[row].contains_key(&val) {
                    count = *rows[row].get(&val).unwrap_or(&0);
                }
                rows[row].insert(val, count+1);
                
                count = 0;
                if cols[col].contains_key(&val) {
                    count = *cols[col].get(&val).unwrap_or(&0);
                }
                cols[col].insert(val, count+1);
                
                count = 0;
                if mini_squares[row/3][col/3].contains_key(&val) {
                    count = *mini_squares[row/3][col/3].get(&val).unwrap_or(&0);
                }
                mini_squares[row/3][col/3].insert(val, count+1);
                
            }
        }
        
        for row in 0..dimension {
            for col in 0..dimension {
                
                for (key, value) in &rows[row] {
                    if *value > 1 && *key != '.' {
                        return false;
                    }
                }
                
                for (key, value) in &cols[col] {
                    if *value > 1 && *key != '.' {
                        return false;
                    }
                }
                
                for (key, value) in &mini_squares[row/3][col/3] {
                    if *value > 1 && *key != '.' {
                        return false;
                    }
                }
                
            }
        }
        
        return true;
    }
}